Feature: Automatic diagram layout

thibaudlabat opened the pull request:
I added a method to the `DefaultPlugin` class to trigger automatic layout for the diagram using `elkjs`.

I also created a Cypress  step definition to test HTML attribute values (may be useful to others).

Zorin95670 commented on package.json:
```diff
@@ -59,6 +59,8 @@
   "dependencies": {
```
Rename commit `Add dependencies: elkjs, web-worker` to `Add dependencies: elkjs, nunjucks, web-worker`

thibaudlabat replied:
I have not added `nunjucks` as a dependency.
See [https://github.com/ditrit/leto-modelizer-plugin-core/blob/main/package.json](https://github.com/ditrit/leto-modelizer-plugin-core/blob/main/package.json)

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ * It has tree-related fields : parent, children, depth,
+ * in addition to the fields from {@link Component} in {@link raw}.
+ * @typedef {object} NodeData
+ * @property {NodeData|null} parent - The node's parent
```
remove |null and change `parent` to `[parent]`

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ * in addition to the fields from {@link Component} in {@link raw}.
+ * @typedef {object} NodeData
+ * @property {NodeData|null} parent - The node's parent
+ * @property {Component|null} raw The underlying component
```
remove |null and change `raw` to `[raw] -`

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ * @typedef {object} NodeData
+ * @property {NodeData|null} parent - The node's parent
+ * @property {Component|null} raw The underlying component
+ * @property {NodeData[]} children This node's children
```
missing `-` after children

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ * @property {NodeData|null} parent - The node's parent
+ * @property {Component|null} raw The underlying component
+ * @property {NodeData[]} children This node's children
+ * @property {number} depth distance from the root (root's depth is 0)
```
missing `-` after depth

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ * Parameters for the ELK automatic layout system.
+ * To be used within {@link AutomaticLayout}.
+ * @typedef {object} ElkParams
+ * @property {boolean} interactive whether we use the interactive mode
```
for all documentation property of ElkParams: missing `-` after property name.

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+import ELK from 'elkjs';
```
for all documentation, missing `.` at the end of each sentences.

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ */
+
+/**
+ * Parameters for the ELK automatic layout system.
```
add `@see` with link to elk documentation of this object.

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ * An ELK port.
+ * (external)
+ * @typedef {object} ElkPort
+ * @see elk-api.d.ts
```
see section is for link, so complete it with correct elk documentation link for all `@see`

Zorin95670 replied:
Or not sure that this typedef is relevant because. So update its documentation or remove it.

thibaudlabat replied:
removed

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+
+// eslint-disable-next-line jsdoc/require-property
+/**
+ * An ELK edge.
```
see section is for link, so complete it with correct elk documentation link for all @see.
Or not sure that this typedef is relevant because. So update its documentation or remove it.

thibaudlabat replied:
removed

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ */
+
+/**
+ * An ELK node.
```
same as ELK edge or ELK port.

thibaudlabat replied:
I added a link to the documentation.

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ * https://github.com/kieler/elkjs
+ *
+ */
+
```
Remove this empty line

thibaudlabat replied:
done

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ * Uses among other things Eclipse Layout Kernel (aka. ELK)
+ *
+ * ELK's documentation is available at:
+ * https://eclipse.dev/elk/
```
use @see for links

thibaudlabat replied:
I used `@link` instead, is it okay ?
`@see` **cannot** be used in the middle of a paragraph.

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+   * @private
+   */
+  __findParentId(node) {
+    // eslint-disable-next-line no-restricted-syntax
```
remove this eslint-disable and use Array.find

thibaudlabat replied:
this method has been deleted

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+  /**
+   * Find the id a node's parent if it has one.
+   * @param {NodeData} node a component data
+   * @returns {string|null} node's parent id or null
```
remove |null

Zorin95670 commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,406 @@
+ *
+ */
+
+// TODO : all methods could be static ; do we need a class or a module is enough ?
```
Agree, remove class and only export function auto.

thibaudlabat replied:
If I do so, then how can I test functions that are not exported ?

I didn't realize that when I wrote this TODO.
Having a class is really useful for Jest's mocking features.

thibaudlabat commented:
Done in the last force-push : 

- updated dependencies
- new DefaultLayout class
- ElkLayout inherits from DefaultLayout (this is where the actual code lives)
- Renamed x,t,e (no more short variable names)
- Renamed `auto` methods to `arrangeComponentsPosition`
- More docs & tests

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+
+    // constructing our data objects from the given components
+    // eslint-disable-next-line no-restricted-syntax
+    for (const c of components) {
```
what is `c` ? please rename this variable

thibaudlabat replied:
changed.

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+   * @returns {Promise<ElkNode>} a layout for the children of {@link parentNode}.
+   * @private
+   */
+  async __generateNodeLayout(parentNode, nodes, links) {
```
rename it as `__generateELKGraph`

thibaudlabat replied:
done

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+   * @returns {Promise<ElkNode[]>} A layout for the graph, see {@link writeLayout}.
+   * @public
+   */
+  async generateLayout(components, links) {
```
rename to `__generateElkGraph`

thibaudlabat replied:
done

thibaudlabat commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,387 @@
+ */
+
+// TODO : all methods could be static ; do we need a class or a module is enough ?
+export class AutomaticLayout {
```
done

thibaudlabat commented on src/draw/AutomaticLayout.js:
```diff
@@ -0,0 +1,387 @@
+   * @constant
+   */
+  static __elk = new ELK();
+
```
(outdated and the referenced source code does not match anymore ? or I don't get it)

thibaudlabat commented on src/draw/DefaultLayout.js:
```diff
@@ -0,0 +1,30 @@
+/**
```
ok

thibaudlabat commented on src/draw/DefaultLayout.js:
```diff
@@ -0,0 +1,30 @@
+   */
+  async arrangeComponentsPosition() {
+    // resolves immediatly and returns nothing
+    return Promise.resolve(undefined);
```
done

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+import ELK from 'elkjs';
```
done

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+    // eslint-disable-next-line no-restricted-syntax
+    for (const c of components) {
+      nodes.set(c.id, {
+        raw: c, children: [], parent: undefined, depth: 0,
```
done

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+    // defines the root
+    /** @type {NodeData} */
+    const root = {
+      raw: undefined, children: [], parent: undefined, depth: 0,
```
done

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+    // the nods without a parent are now children of the root
+    // eslint-disable-next-line no-restricted-syntax
+    for (const node of nodes.values()) {
+      if (!node.parent) {
```
done

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+    // computes depth for each node
+    // eslint-disable-next-line no-restricted-syntax
+    for (const node of nodes.values()) {
+      node.depth = this.__computeNodeDepth(node);
```
changed

thibaudlabat commented:
Since my last comment :

- package updates (again)
- deleted __findParentId, using instead getContainerId
- deleted big json files / functional tests
- `for...of` --> `L.forEach`
- reorder methods in ElkLayout
+ changes according to the above comments

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,344 @@
+   * @returns {number} depth.
+   * @private
+   */
+  __computeNodeDepth(node) {
```
```js
getNodeDepth(node) {
  const depth = 0;
  let currentNode = node.parent;
  
  while(parent) {
    depth += 1;
    currentNode = currentNode.parent;
  }
  
  return depth;
}
```

DimitriChauvel replied:
done

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,344 @@
+
+    return allLinks
+    // only links with valid source/target
+      .filter((link) => nodes.get(link.source) !== undefined
```
remove this, can't have link without source or target.

DimitriChauvel replied:
done

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,344 @@
+        return [sourceDepth, targetDepth];
+      })
+    // source and target are different and non-falsy
+      .filter((link) => link[0] !== link[1] && link[0] && link[1])
```
remove this

DimitriChauvel replied:
done

thibaudlabat replied:
undone; the check was in fact useful. After removing it and running the tests, we have a self-link from `workflow1` to `workflow1`, which is not wanted.

thibaudlabat replied:
Replaced with `.filter((link) => link[0] !== link[1])`

thibaudlabat replied:
(I think I deleted Dimitri comment by mistake, wanting to delete mine)
(It said : Done)

thibaudlabat replied:
Why would we remove this ? It makes the E2E tests crash and without it the demo does not run anymore.

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,344 @@
+   * @returns {[string,string][]} inner links represented as pairs of nodes ids.
+   * @private
+   */
+  __getLinksForChildren(nodes, allLinks, parentNode) {
```
why not something like this: 

```js
getChildrenLinks(parentNode, links) {
    const childrenIds = new Set(parentNode.children.map((node) => node.raw?.id));
    return links
      .filter(({ source, target }) => childrenIds.has(source) && childrenIds.has(target))
      .map(({ source, target }) => ({
        id: `${source}__${target}`, sources: [source], targets: [target],
      }));
}
```

thibaudlabat replied:
I essentially wrote the same code, but taking into account the hierarchy (which part you removed).

 Hierarchical rendering (depth-first) allows to compute the correct size for the inner components of each parent component, triggering an ELK repositioning each time.

What you wrote would mean that we provide to ELK links that are not at the same depth in the tree, or that do not have the same parent, which is not possible at the time.

That's why I used depth / `__getAncestorByDepth` (used to be `__climbToDepth`), to get links at the same level in the tree.

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,344 @@
+   * @returns {NodeData} `from` 's ancestor or null.
+   * @private
+   */
+  __climbToDepth(from, depth) {
```
rename to `getParentNodeByDepth`

DimitriChauvel replied:
this funtion name "already" exists.

thibaudlabat replied:
renamed to `__getAncestorByDepth`

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,344 @@
+    // defines the root
+    /** @type {NodeData} */
+    const root = {
+      raw: null, children: [], parent: null, depth: 0,
```
Instead of having an array for children, why not directly use a map ?

thibaudlabat replied:
I assume that you are talking about using a `Map` object for key-value relationship.

What's the point ? 

We are always iterating over all children and never looking for a specific child. (which is why we would use a `Map` object) 
And `Map` objects have no `.map( ...)` method to help for iteration.

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+  static __elk = new ELK();
+
+  /**
+   * @param {DefaultData} pluginData - a graph to be arranged.
```
done

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+ *
+ * What we use here elkjs, which is a transpiled version
+ * from Java to Javascript :
+ * {@link https://github.com/kieler/elkjs}
```
done

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+   */
+  async __generateNodeLayout(parentNode, nodes, links) {
+    // default parameters
+    const effectiveParams = {
```
done

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,379 @@
+
+    // we prepare the input in the format expected by ELK
+    const graph = {
+      id: parentNode.raw ? parentNode.raw.id : 'root', layoutOptions, children: [], edges: [],
```
done

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,377 @@
+ */
+
+/**
+ * @augments DefaultLayout
```
move at the end of documentation

thibaudlabat replied:
done

Zorin95670 commented on tests/unit/draw/DefaultLayout.spec.js:
```diff
@@ -0,0 +1,19 @@
+
+describe('Test Class: DefaultLayout', () => {
+  describe('Test constructor', () => {
+    it('correctly defines this.pluginData', () => {
```
change by:
```js
  describe('Test constructor', () => {
    it('Checks variable instantiation', () => {
      const layout = new DefaultLayout();

      expect(layout.pluginData).toBeNull();
    });

    it('Checks passing variable to constructor', () => {
      const layout = new DefaultLayout(new DefaultData());

      expect(layout.pluginData).toEqual(new DefaultData());
    });
  });
```

thibaudlabat replied:
done

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,377 @@
+  /**
+   * Initializes ELK parameters and inherited fields.
+   * @param {DefaultData} pluginData - a graph to be arranged.
+   * @param {object} [elkParams] - Parameters for the layout algorithm. (use defaults if unsure)
```
add @link to elk parameters documentation.

thibaudlabat replied:
done

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,377 @@
+   */
+  writeSingleDepthLayout(elkNode) {
+    const nodes = new Map(elkNode.children
+      .map((node) => [node.id, {
```
```js
.map((node) => [node.id, {
  x: node.x,
  y: node.y,
}]));
```

thibaudlabat replied:
done

Zorin95670 commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,377 @@
+        x: node.x, y: node.y,
+      }]));
+
+    this.pluginData.components.forEach((component) => {
```
```js
this.pluginData.components.
  .filter(({ id }) => node.has(id))
  .forEach(...)
```

thibaudlabat replied:
done

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+  describe('Test method: getNodes', () => {
+    const layoutManager = new ElkLayout(undefined);
+
+    it('should throw exception on undefined argument', () => {
```
Useless test, because components can't be undefined. Always equals to an array.

thibaudlabat replied:
removed

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+      const node = nodes.get('my_id');
+
+      expect(node).toBeDefined();
+      expect(node?.raw).toBe(component);
```
remove `?`

thibaudlabat replied:
removed

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+
+      expect(node).toBeDefined();
+      expect(node?.raw).toBe(component);
+
```
remove this line and other between expect

thibaudlabat replied:
done

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+      expect(node.depth).toEqual(1);
+    });
+
+    it('there is a (single) root', () => {
```
Rework on test description to explain what you test and what you expect

thibaudlabat replied:
done

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+      expect(root.parent).toBeNull();
+    });
+
+    it('captures parent-child relationship and depth in the expected way', () => {
```
should

thibaudlabat replied:
done

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+      expect(node3.depth).toEqual(3);
+    });
+
+    // TODO : error on missing parent / link ? (do we want an error ?)
```
remove this

thibaudlabat replied:
removed

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+  describe('Test method: getNodeDepth', () => {
+    const layoutManager = new ElkLayout();
+
+    it('computes depth correctly', () => {
```
should

thibaudlabat replied:
done

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+    });
+
+    it('has expected behavior with 3 given components', () => {
+      const comp1 = new Component({
```
renomae to child, parent and ancestor

thibaudlabat replied:
done

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+  });
+
+  describe('Test method: generateLayout', () => {
+    const layoutManager = new ElkLayout(undefined);
```
remove undefined

thibaudlabat replied:
removed

Zorin95670 commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+  });
+
+  describe('Test method: getNodeDepth', () => {
+    const layoutManager = new ElkLayout();
```
for all, declere layout in first line of `it` to be sure that layout is not impacted by other test.

thibaudlabat replied:
done

Zorin95670 commented on cypress/support/step_definitions/html.js:
```diff
@@ -124,3 +126,11 @@ Then('I expect {string} appear {int} time(s) on screen', (templateSelector, coun
   cy.get(selector).should('have.length', count);
 });
+
+Then('I expect {string} attribute {string} to be {string}', (templateSelector, attribute, value) => {
```
Remote this because is not used.

thibaudlabat replied:
done

Zorin95670 commented on cypress/support/step_definitions/html.js:
```diff
@@ -126,11 +126,3 @@ Then('I expect {string} appear {int} time(s) on screen', (templateSelector, coun
   cy.get(selector).should('have.length', count);
 });
-
-Then('I expect {string} attribute {string} to be {string}', (templateSelector, attribute, value) => {
```
Misplaced deletion

thibaudlabat replied:
done

Zorin95670 commented on cypress/e2e/automatic-layout.feature:
```diff
@@ -0,0 +1,30 @@
+
+    And I expect "#server1" to be at position 12,12
+    And I expect "#server2" to be at position 12,112
+
```
remove one empty line

Zorin95670 commented on changelog.md:
```diff
@@ -14,6 +14,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - Add `setId`, `createAttribute` and `getAttributesByType` inside `Component.js`.
   - Add `renameComponentId` inside `DefaultData.js`.
 
+- Add `arrangeComponentsPosition` method in `DefaultDrawer` and `DefaultPlugin`
```
You miss to add `guides/migrations/0.18.0_to_0.X.0.md`
And show change in DefaultDrawer. You can read previous migration guides for example.

thibaudlabat replied:
What kind of change is expected from whoever uses DefautPlugin/DefaultDrawer ? I add a feature and do not modify anything that could be previously used.

thibaudlabat replied:
done

thibaudlabat commented on tests/unit/draw/ElkLayout.spec.js:
```diff
@@ -0,0 +1,494 @@
+      const node1 = nodes.get('my_id_1');
+      const node2 = nodes.get('my_id_2');
+
+      expect(node1?.raw).toBe(comp1);
```
done

thibaudlabat commented on tests/unit/draw/DefaultDrawer.spec.js:
```diff
@@ -1082,4 +1083,22 @@ describe('Test Class: DefaultDrawer()', () => {
+      'root',
+    );
+
+    it('propagates arrangeComponentsPosition call to this.layout', async () => {
```
done

thibaudlabat commented on tests/unit/models/DefaultPlugin.spec.js:
```diff
@@ -257,4 +258,18 @@ describe('Test class: DefaultPlugin', () => {
+      },
+    });
+
+    it('propagates arrangeComponentsPosition call to this.__drawer', async () => {
```
done

Zorin95670 commented on cypress/support/step_definitions/html.js:
```diff
@@ -98,10 +98,12 @@ Then('I expect {string} to have {string} as parent', (templateSelector, template
+    expect(Math.trunc(parseInt(element.attr('x')))).eq(x);
+    expect(Math.trunc(parseInt(element.attr('y')))).eq(y);
   });
+
```
remove this empty line

thibaudlabat replied:
done

k-y0u commented on src/draw/DefaultDrawer.js:
```diff
@@ -21,13 +22,20 @@ class DefaultDrawer {
+   * @param {DefaultLayout} [layout] - The manager for an automatic diagram layout
    */
-  constructor(pluginData, resources = null, rootId = 'root', options = {}) {
+  constructor(pluginData, resources = null, rootId = 'root', options = {}, layout = undefined) {
```
`undefined` is already the default value, so no need to specify it.

thibaudlabat replied:
done

thibaudlabat replied:
We cannot have a non-optional argument (ie. without an explicit default value) after an optional argument.
(`ESLint: Default parameters should be last.(default-param-last)`)
Moving `layout` as the second argument would be a breaking change.

I suggest to leave it as it is.
Do you agree ?

Zorin95670 replied:
Set `null` instead as`resources`

thibaudlabat replied:
done

justinehell commented on src/draw/DefaultDrawer.js:
```diff
@@ -21,13 +22,20 @@ class DefaultDrawer {
    * @param {number[]} [options.lineLengthPerDepth] - Number of components
    * per line at a given depth. Valid values: 1 - Infinity.
    * @param {number} [options.actionMenuButtonSize] - The size of each action menu button.
+   * @param {DefaultLayout} [layout] - The manager for an automatic diagram layout
```
missing .

thibaudlabat commented on src/draw/ElkLayout.js:
```diff
@@ -0,0 +1,380 @@
+
+  /**
+   * Write a generated layout to the corresponding graph.
+   * @param {ElkNode[]} layout - previously generated layout, see {@link generateAllElkLayouts}.
```
I added capital letters everywhere.

thibaudlabat commented on src/draw/DefaultDrawer.js:
```diff
@@ -1723,5 +1731,15 @@ class DefaultDrawer {
+  /**
+   * Reorganize nodes layout algorithmically.
+   * This method does not refresh the view.
+   * You have to await it and trigger a redraw
```
done

thibaudlabat commented on changelog.md:
```diff
@@ -14,6 +14,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - Add `renameComponentId` inside `DefaultData.js`.
 
+- Add `arrangeComponentsPosition` method in `DefaultDrawer` and `DefaultPlugin`
+- Add inheritable class `DefaultLayout` for automatic layout in the diagram (does nothing)
+- Add inherited class `ElkLayout` for automatic layout in the diagram, using ELK.
+- Add HTML attribute equality as a Cypress step definition in `html.js`
+- Add `elkjs` and `web-worker` as package dependencies
```
done
